log_cache_test.go

```diff
@@ -86,3 +88,56 @@ func TestLogCache(t *testing.T) {
 		t.Fatalf("err: %v", err)
 	}
 }
+
+type errorStore struct {
+	LogStore
+	mu      sync.Mutex
+	fail    bool
+	failed  int
+	failMax int
+}
+
+func (e *errorStore) StoreLogs(logs []*Log) error {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	if e.fail {
+		e.failed++
+		if e.failed <= e.failMax {
+			return errors.New("some error")
+		}
+		e.fail = false
+	}
+	return e.LogStore.StoreLogs(logs)
+}
+
+func (e *errorStore) failNext(count int) {
+	e.mu.Lock()
+	e.fail = true
+	e.failMax = count
+	e.mu.Unlock()
+}
+
+func TestLogCacheWithBackendStoreError(t *testing.T) {
+	store := NewInmemStore()
+	errStore := &errorStore{LogStore: store}
+	c, _ := NewLogCache(16, errStore)
+
+	for i := 0; i < 4; i++ {
+		log := &Log{Index: uint64(i) + 1}
+		store.StoreLog(log)
+	}
+	errStore.failNext(1)
+	log := &Log{Index: 5}
+	c.StoreLog(log)
+
+	var out Log
+	for i := 0; i < 4; i++ {
+		if err := c.GetLog(uint64(i+1), &out); err != nil {
```

Comment:
I got a bit confused following the indexes and seeing the `i+1` here.  How does looping from 1 to 5 instead to avoid the index arithmetic sound?